src/psyclone/psyad/domain/lfric/lfric_adjoint_harness.py

```diff
@@ -365,32 +426,70 @@ def generate_lfric_adjoint_harness(tl_psyir):
     kern = lfalg.kernel_from_metadata(parse_tree, kernel_name)
     kern_args = lfalg.construct_kernel_args(routine, kern)
 
+    # Validate the index values for the coordinate and face_id fields if
+    # supplied.
+    geometry_arg_indices = []
+    if coord_arg_idx is not None:
+        _validate_geom_arg(kern, coord_arg_idx, "coordinate", ["wchi"], 3)
+        # Convert to 0-indexed
+        coord_arg_idx -= 1
+        geometry_arg_indices.append(coord_arg_idx)
+    if panel_id_arg_idx is not None:
+        # Check that the specified argument is of the correct type.
+        _validate_geom_arg(kern, panel_id_arg_idx, "panel-id",
+                           LFRicConstants().VALID_DISCONTINUOUS_NAMES, 1)
+        # Convert to 0-indexed
+        panel_id_arg_idx -= 1
+        geometry_arg_indices.append(panel_id_arg_idx)
+
     # Create symbols that will store copies of the inputs to the TL kernel.
     # Currently we only support scalar and field arguments.
+    field_args = [fsym for fsym, _ in kern_args.fields]
+    scalar_and_field_args = kern_args.scalars + field_args
+
     # TODO #1864 - add support for operators.
     input_symbols = {}
-    for sym in kern_args.scalars + [fsym for fsym, _ in kern_args.fields]:
+    for sym in scalar_and_field_args:
+        idx = kern_args.arglist.index(sym.name)
+        if (kern_args.metadata_index_from_actual_index(idx) in
+                geometry_arg_indices):
+            # This kernel argument is not modified by the test harness so we
+            # don't need to keep a copy of it.
+            continue
         input_sym = table.new_symbol(sym.name+"_input", symbol_type=DataSymbol,
                                      datatype=DeferredType())
         input_sym.copy_properties(sym)
         input_symbols[sym.name] = input_sym
 
-    # Initialise all input field objects.
+    # Initialise all input field objects unless they contain geometric
+    # information (as specified by the user via command-line arguments).
+    fld_arg_list = []
```

Comment:
The list name and its use is confusing (despite the good comments). Could you rename. Sorry I have no good suggestion :-)

Reply:
I've gone for `kernel_input_arg_list`. It's longer so it must be better ;-)